core/src/main/java/org/apache/calcite/rex/RexUtil.java

```diff
@@ -2833,4 +2898,77 @@ public boolean anyContain(Iterable<? extends RexNode> nodes) {
     }
   }
 
+  /** Converts a {@link Range} to a {@link RexNode} expression.
+   *
+   * @param <C> Value type */
+  private static class RangeToRex<C extends Comparable<C>>
+      implements RangeSets.Consumer<C> {
+    private final List<RexNode> list;
+    private final RexBuilder rexBuilder;
+    private final RelDataType type;
+    private final SqlTypeName typeName;
+    private final RexNode ref;
```

Comment:
They didn't write `'foo '`. They created a Sarg with `'buzz'` and `'foo '`. The `Sarg` converted both values to `CHAR(4)`.

Spaces on the ends of `CHAR` literals can literally be ignored.

Reply:
> Spaces on the ends of CHAR literals can literally be ignored.

I kind of remember this seems a SQL standard, what i'm curious of is should we do the trim specifically for `Sarg` here ? What about a literal without `Sarg`, there was no trim before this patch.

Reply:
Sargs are the only case (thus far) where we combine several literals of similar types into a collection all with the widest type, and later want to separate them into literals with their own, narrower, types.

Reply:
Is there possibility that we record the real literal types somewhere in the `Sarg` structure so that when we need them there is no need for such a magic `trim` conversion.

Reply:
Yes, but that is an inferior solution. Sargs are good because they are a more concise and uniform representation of certain kinds of expressions. The information that they remove was redundant. Transforming CHAR(3) 'foo' to CHAR(4) 'foo ' is lossless and therefore reversible. Let's embrace that.

Reply:
> Spaces on the ends of CHAR literals can literally be ignored

I'm afraid that not every db engine follow the rule. Here is my testing:

- MySQL 5.4 and 8.0 true
- PostgreSQL 9.6 and 12 false
- SQLite 3.3.0 false

Reply:
I'm still not convinced there is a problem. We are doing these transformations in Calcite's algebra and semantics, not PostgreSQL's.

Reply:
If a downstream project have a `Sarg` expression `f0 in ('foo ', 'bar', 'zoo')`, before this patch, the `'foo '` will never be trimmed, is this still true when we translate `Sarg` to `RexNode` ? If it is not, that would probably a breaking change because not every engine follow the "Spaces on the ends of CHAR literals can literally be ignored" rule.

Reply:
But that's the downstream project's problem, right? They sent the algebra to Calcite. They get Calcite semantics.

I'm trying to create a simple abstraction here. You are making it more complex. If it doesn't meet your requirements and is useless to you, turn it off. But if we make the abstraction more complicated, it is now useless to everyone.

In your example `f0 in ('foo ', 'bar', 'zoo')` if the types `CHAR(4)`, `CHAR(3)` and `CHAR(3)` are not compatible in your type system, then this is not a candidate for Sargs. I don't know the details, but you might find that in PostgreSQL the types are actually `VARCHAR(4)`, `VARCHAR(3)` and `VARCHAR(3)`. You figure it out. I'm not changing Sarg to accommodate.